lighthouse-core/audits/polyfills.js

```diff
@@ -0,0 +1,225 @@
+      'String.prototype.repeat',
+      'String.prototype.startsWith',
+      'String.prototype.endsWith',
+      'String.prototype.padStart',
+      'String.prototype.padEnd',
+      'String.prototype.trim',
+      'Function.prototype.bind',
+      'Node.prototype.append',
+      'Node.prototype.prepend',
+      'Node.prototype.before',
+      'Node.prototype.after',
+      'Node.prototype.remove',
+      'Node.prototype.replaceWith',
+      'Node.prototype.children',
+      'NodeList.prototype.forEach',
+      'Element.prototype.closest',
+      'Element.prototype.toggleAttribute',
+      'Element.prototype.matches',
+      'Element.prototype.classList',
+      'MouseEvent',
+      'CustomEvent',
+      'Number.isNaN',
+      'HTMLCanvasElement.prototype.toBlob',
+    ].map(str => {
+      const parts = str.split('.');
+      return {
+        object: parts.length > 1 ?
+                parts.slice(0, parts.length - 1).join('.')
+          : undefined,
+        property: parts[parts.length - 1],
+      };
+    });
+
+    /** @type {Map<Poly, number>} */
+    const polyCounter = new Map();
+
+    /** @type {Map<PolyIssue, number>} */
+    const polyIssueCounter = new Map();
+
+    /** @type {Map<string, PolyIssue[]>} */
+    const urlToPolyIssues = new Map();
+
+    for (const [requestId, content] of Object.entries(artifacts.Scripts)) {
+      const networkRecord = networkRecords.find(record => record.requestId === requestId);
+      if (!networkRecord) continue;
+      const extPolys = this.detectPolys(polys, content);
+      urlToPolyIssues.set(networkRecord.url, extPolys);
+      for (const polyIssue of extPolys) {
+        const val = polyCounter.get(polyIssue.poly) || 0;
+        polyIssueCounter.set(polyIssue, val);
+        polyCounter.set(polyIssue.poly, val + 1);
+      }
+    }
+
+    /** @type {Array<{url: string, description: string, location: string}>} */
+    const tableRows = [];
+    urlToPolyIssues.forEach((polyIssues, url) => {
+      for (const polyIssue of polyIssues) {
+        const {poly, row, col} = polyIssue;
+        const polyStatement = `${poly.object ? poly.object + '.' : ''}${poly.property}`;
+        const numOfThisPoly = polyCounter.get(poly) || 0;
+        const polyIssueOccurrence = polyIssueCounter.get(polyIssue) || 0;
+        const isMoreThanOne = (polyCounter.get(poly) || 0) > 1;
+        const countText = isMoreThanOne ? ` (${polyIssueOccurrence + 1} / ${numOfThisPoly})` : '';
+        tableRows.push({
+          url,
+          description: `${polyStatement}${countText}`,
+          // TODO use sourcemaps
+          location: `r: ${row}, c: ${col}`,
+        });
+      }
+    });
+    const headings = [
+      {key: 'url', itemType: 'url', text: 'URL'},
+      {key: 'description', itemType: 'code', text: 'Description'},
+      {key: 'location', itemType: 'code', text: 'Location'},
+      // TODO include estimate for size based on https://www.npmjs.com/package/mdn-polyfills#supported-polyfills ?
```

Comment:
this is interesting, we could then make it an opportunity instead of a diagnostic too!

Reply:
I think it might be hilariously small of an impact, though. The average in that table is 300 bytes. Say there's a dozen - about 4KB. is that significant?

Reply:
We'd probably have to get a little more aggressive then. Some of the babel polyfills are no joke, maybe async/await IIRC, but the impact was more like ~80KB for some bundles.

Reply:
Yeah in phil's article it was ~95K minified plus some extra parse/eval time savings

![image](https://user-images.githubusercontent.com/2301202/49826693-26cbce00-fd4d-11e8-809e-12c57e3db437.png)


https://philipwalton.com/articles/deploying-es2015-code-in-production-today/

Reply:
But do we know how much of that is generator cruft (async -> es5 code), string templates (aka language level polyfills), vs. web api polyfills?

so far I'm just considering web api polyfills, which is much more easily identifiable than language polyfills. It'd be very powerful to capture those too though. any ideas?

Reply:
Yeah these are great points!

I wasn't trying to suggest we do it right away these other categories are what I was trying to get at with be more aggressive :) I'm afraid I don't really have any obvious initial ideas we can pursue. Maybe we'll have to take a look at some babelized bundles and look for distinctive markers of cruft, but we can cross that bridge another day for sure 👍